Compression/Huffman.js

```diff
@@ -0,0 +1,146 @@
+  const freqTable = {}
+
+  for (const char of data) {
+    freqTable[char] = (freqTable[char] || 0) + 1
+  }
+
+  return freqTable
+}
+
+/**
+ * A Huffman Node is a node in a Huffman tree.
+ * @class HuffmanNode
+ * @property {string} char - The character represented by the node.
+ * @property {number} freq - The frequency of the character.
+ * @property {HuffmanNode} left - The left child of the node.
+ * @property {HuffmanNode} right - The right child of the node.
+ */
+class HuffmanNode {
+  constructor(char, freq) {
+    this.char = char
+    this.freq = freq
+    this.left = null
+    this.right = null
+  }
+}
+
+/**
+ * Builds a Huffman tree from a frequency table.
+ * @param {Object} freqTable - The frequency table to use for building the tree.
+ * @returns {HuffmanNode} - The root node of the Huffman tree.
+ */
+function buildHuffmanTree(freqTable) {
+  const nodes = Object.keys(freqTable).map(
+    (char) => new HuffmanNode(char, freqTable[char])
+  )
+
+  while (nodes.length > 1) {
+    nodes.sort((a, b) => a.freq - b.freq)
+    const left = nodes.shift()
+    const right = nodes.shift()
+    const parent = new HuffmanNode(null, left.freq + right.freq)
+    parent.left = left
+    parent.right = right
+    nodes.push(parent)
+  }
+
+  return nodes[0]
+}
+
+/**
+ * Builds a Huffman code table from a Huffman tree.
+ * @param {HuffmanNode} root - The root node of the Huffman tree.
+ * @param {string} [prefix=''] - The prefix to use for the Huffman codes.
+ * @param {Object} [codes={}] - The Huffman code table.
+ * @returns {Object} - The Huffman code table.
+ */
+function buildHuffmanCodes(root, prefix = '', codes = {}) {
+  if (root) {
+    if (root.char) {
+      codes[root.char] = prefix
+    }
+    buildHuffmanCodes(root.left, prefix + '0', codes)
+    buildHuffmanCodes(root.right, prefix + '1', codes)
+  }
+  return codes
+}
+
+/**
+ * Encodes a string using Huffman Coding.
+ * @param {string} data - The string to encode.
+ * @param {Object} freqTable - The frequency table to use for encoding.
+ * @returns {string} - The encoded string.
+ */
+function encodeHuffman(data, freqTable) {
+  const root = buildHuffmanTree(freqTable)
+  const codes = buildHuffmanCodes(root)
+
+  let encodedData = ''
+  for (let char of data) {
+    encodedData += codes[char]
+  }
+
+  return encodedData
+}
+
+/**
+ * Decodes a string using Huffman Coding.
+ * @param {string} encodedData - The string to decode.
+ * @param {HuffmanNode} root - The root node of the Huffman tree.
+ * @returns {string} - The decoded string.
+ */
+function decodeHuffman(encodedData, root) {
+  let decodedData = ''
+  let currentNode = root
+  for (let bit of encodedData) {
+    if (bit === '0') {
+      currentNode = currentNode.left
+    } else {
+      currentNode = currentNode.right
+    }
+
+    if (currentNode.char) {
+      decodedData += currentNode.char
+      currentNode = root
+    }
+  }
+
+  return decodedData
+}
+
+// Example usage
```

Comment:
This does not belong here. It should be in the tests (where it already is) / JSDoc examples (where some methods could still use some of it).